README.md
# http_signature_helper  [![pub package](https://img.shields.io/pub/v/http_signature_helper.svg)](https://pub.dartlang.org/packages/http_signature_helper)

Signature helpers from the [HTTP Signature spec](https://tools.ietf.org/id/draft-cavage-http-signatures-10.html).


## Usage

```dart
import 'package:http_signature_helper/http_signature_helper.dart';

main() {
  final signatureString = SignatureString(
    body: "body",
    headers: {
      "Host": "example.org",
      "Date": "Tue, 07 Jun 2014 20:51:35 GMT",
      "X-Example": """Example header
    with some whitespace.""",
      "Cache-Control": "max-age=60, must-revalidate"
    },
    signatureHeaders: [
      "(request-target)",
      "host",
      "date",
      "cache-control",
      "x-example"
    ],
    target: SignatureTarget("GET", "/foo"),
  );

  print(signatureString);
  // (request-target): get /foo\nhost: example.org\ndate: Tue, 07 Jun 2014 20:51:35 GMT\ncache-control: max-age=60, must-revalidate\nx-example: Example header with some whitespace.\nbody

  final signatureHeader = SignatureHeader(
    keyId: "rsa-key-1",
    algorithm: "rsa-sha256",
    signatureHeaders: [
      "(request-target)",
      "host",
      "date",
      "digest",
      "content-length"
    ],
    signature: "rsa-signature-1",
  );

  print(signatureHeader);
  // Signature keyId="rsa-key-1",algorithm="rsa-sha256",headers="(request-target) host date digest content-length",signature="rsa-signature-1"

  final signatureHeader2 = SignatureHeader.parse(signatureHeader.toString());

  print(signatureHeader == signatureHeader2);
  // true
}

```

## Features and bugs

Please file feature requests and bugs at the [issue tracker][tracker].

[tracker]: https://github.com/Cretezy/http_signature_helpers.dart/issues


CHANGELOG.md
## 1.0.1 (2018-08-14)

- Add defaults for headers

## 1.0.0 (2018-08-12)

- Initial version


test/http_signature_helper_test.dart
import 'package:http_signature_helper/http_signature_helper.dart';
import 'package:test/test.dart';

void main() {
  test('create signature string', () {
    final correct =
        "(request-target): get /foo\nhost: example.org\ndate: Tue, 07 Jun 2014 20:51:35 GMT\ncache-control: max-age=60, must-revalidate\nx-example: Example header with some whitespace.\nbody";

    final signatureString = SignatureString(
      body: "body",
      headers: {
        "Host": "example.org",
        "Date": "Tue, 07 Jun 2014 20:51:35 GMT",
        "X-Example": """Example header
    with some whitespace.""",
        "Cache-Control": "max-age=60, must-revalidate"
      },
      signatureHeaders: [
        "(request-target)",
        "host",
        "date",
        "cache-control",
        "x-example"
      ],
      target: SignatureTarget("GET", "/foo"),
    );

    expect(signatureString.toString(), correct);
  });

  test("creates signature header", () {
    final correct =
        'Signature keyId="rsa-key-1",algorithm="rsa-sha256",headers="(request-target) host date digest content-length",signature="rsa-signature-1"';
    final signatureHeader = SignatureHeader(
      keyId: "rsa-key-1",
      algorithm: "rsa-sha256",
      signatureHeaders: [
        "(request-target)",
        "host",
        "date",
        "digest",
        "content-length"
      ],
      signature: "rsa-signature-1",
    );

    expect(signatureHeader.toString(), correct);
  });

  test("parses signature header", () {
    final correct = SignatureHeader(
      keyId: "rsa-key-1",
      algorithm: "rsa-sha256",
      signatureHeaders: [
        "(request-target)",
        "host",
        "date",
        "digest",
        "content-length"
      ],
      signature: "rsa-signature-1",
    );
    final check = SignatureHeader.parse(
      'Signature keyId="rsa-key-1",algorithm="rsa-sha256",headers="(request-target) host date digest content-length",signature="rsa-signature-1"',
    );

    expect(check, correct);
  });
}


lib/http_signature_helper.dart
library http_signature_helper;

import 'package:auth_params/auth_params.dart';
import 'package:collection/collection.dart';

class SignatureTarget {
  final String method;
  final String path;

  SignatureTarget(this.method, this.path);
}

class SignatureNonce {
  final String clientId;
  final String nonce;

  SignatureNonce(this.clientId, this.nonce);

  @override
  bool operator ==(other) {
    return clientId == other.clientId && nonce == other.nonce;
  }
}

class SignatureString {
  final String body;
  final Map<String, String> headers;
  final List<String> signatureHeaders;
  final SignatureTarget target;
  final SignatureNonce nonce;

  SignatureString(
      {this.body,
      Map<String, String> headers = const <String, String>{},
      List<String> signatureHeaders = const <String>[],
      this.target,
      this.nonce})
      : this.headers =
            headers?.map((key, value) => MapEntry(key.toLowerCase(), value)),
        this.signatureHeaders = signatureHeaders
            ?.map((signatureHeader) => signatureHeader.toLowerCase())
            ?.toList();

  @override
  String toString() {
    final lines = List<String>();

    if (nonce != null) {
      lines.add("${nonce.clientId} ${nonce.nonce}");
    }

    if (signatureHeaders.contains("(request-target)") && target != null) {
      lines.add(
          "(request-target): ${target.method.toLowerCase()} ${target.path}");
    }

    signatureHeaders.forEach((signatureHeader) {
      if (!headers.containsKey(signatureHeader)) {
        // Header is in [signatureHeaders], but not in [headers], skip
        return;
      }
      final header = headers[signatureHeader]
          .split("\n")
          .map((header) => header.trim())
          .join(" ");
      lines.add("$signatureHeader: $header");
    });

    if (body != null) {
      lines.add(body);
    }

    return lines.join("\n");
  }
}

class SignatureHeader {
  final String keyId;
  final String algorithm;
  final String signature;
  final List<String> signatureHeaders;
  final SignatureNonce nonce;
  final bool prefixed;

  SignatureHeader(
      {this.keyId,
      this.algorithm,
      this.signature,
      List<String> signatureHeaders = const <String>[],
      this.nonce,
      this.prefixed = true})
      : this.signatureHeaders = signatureHeaders
            ?.map((signatureHeader) => signatureHeader.toLowerCase())
            ?.toList();

  @override
  String toString() {
    final output = Map<String, String>();

    if (keyId != null) {
      output["keyId"] = keyId;
    }
    if (algorithm != null) {
      output["algorithm"] = algorithm;
    }
    if (signatureHeaders != null) {
      output["headers"] = signatureHeaders.join(" ");
    }
    if (signature != null) {
      output["signature"] = signature;
    }
    if (nonce != null) {
      output["clientId"] = nonce.clientId;
      output["nonce"] = nonce.nonce;
    }

    return (prefixed ? "Signature " : "") + authParams.encode(output);
  }

  factory SignatureHeader.parse(String input, [bool prefixed = true]) {
    if (prefixed && input.startsWith("Signature ")) {
      input = input.replaceFirst("Signature ", "");
    }
    final decoded = authParams.decode(input);

    return SignatureHeader(
      keyId: decoded["keyId"],
      algorithm: decoded["algorithm"],
      signatureHeaders: decoded["headers"]?.split(" "),
      signature: decoded["signature"],
      nonce: decoded["nonce"] != null
          ? SignatureNonce(
              decoded["clientId"],
              decoded["nonce"],
            )
          : null,
    );
  }

  @override
  bool operator ==(other) {
    if (keyId != other.keyId) {
      return false;
    }
    if (algorithm != other.algorithm) {
      return false;
    }
    if (keyId != other.keyId) {
      return false;
    }
    if (signature != other.signature) {
      return false;
    }
    if (!ListEquality().equals(signatureHeaders, other.signatureHeaders)) {
      return false;
    }
    if (nonce != other.nonce) {
      return false;
    }
    if (prefixed != other.prefixed) {
      return false;
    }

    return true;
  }
}


example/http_signature_helper_example.dart
import 'package:http_signature_helper/http_signature_helper.dart';

main() {
  final signatureString = SignatureString(
    body: "body",
    headers: {
      "Host": "example.org",
      "Date": "Tue, 07 Jun 2014 20:51:35 GMT",
      "X-Example": """Example header
    with some whitespace.""",
      "Cache-Control": "max-age=60, must-revalidate"
    },
    signatureHeaders: [
      "(request-target)",
      "host",
      "date",
      "cache-control",
      "x-example"
    ],
    target: SignatureTarget("GET", "/foo"),
  );

  print(signatureString);
  // (request-target): get /foo\nhost: example.org\ndate: Tue, 07 Jun 2014 20:51:35 GMT\ncache-control: max-age=60, must-revalidate\nx-example: Example header with some whitespace.\nbody

  final signatureHeader = SignatureHeader(
    keyId: "rsa-key-1",
    algorithm: "rsa-sha256",
    signatureHeaders: [
      "(request-target)",
      "host",
      "date",
      "digest",
      "content-length"
    ],
    signature: "rsa-signature-1",
  );

  print(signatureHeader);
  // Signature keyId="rsa-key-1",algorithm="rsa-sha256",headers="(request-target) host date digest content-length",signature="rsa-signature-1"

  final signatureHeader2 = SignatureHeader.parse(signatureHeader.toString());

  print(signatureHeader == signatureHeader2);
  // true
}
